File: 005/python/005.opt.py
from functools import reduce

# returns the Greatest Common Divisor for "a" and "b"
def gcd(a, b):

  if b > a:
    return gcd(b, a)
  
  if a % b == 0:
    return b
  
  return gcd(b, a % b)

# returns the Least Common Multiple of "a" and "b"
def lcm(a, b):
  return a * b / gcd(a, b)

# return the GCD or LCM for a list of numbers
def lcm_(*args):
  return reduce(lcm, args)
def gcd_(*args):
  return reduce(gcd, args)

# test
print(lcm_(*range(1, 11)))
print(lcm_(*range(1, 21)))

File: 005/js/005.opt.js
// function for a pair of numbers
/* function gcd(a, b) {
  return (a < b) ? gcd(b, a)
       : (a % b == 0) ? b
       : gcd(b, a % b)
} */

function gcd(...numbers) {
  // returns the greatest common divisor
  // of given numbers
  return numbers.reduce((a, b) => {
    return (a < b) ? gcd(b, a)
         : (a % b == 0) ? b
         : gcd(b, a % b)
  });
}

function lcm(...numbers) {
  // returns the largest common multiple
  // of given numbers
  return numbers.reduce((a, b) => {
    return (a * b) / gcd(a, b);
  });
}


// tests
let numbers = [... Array(20).keys()].slice(1);  // [1, 2, ... 20]
console.log(lcm(...numbers));